docs/Custom_Operation_for_GPUs.md

```diff
@@ -558,1657 +558,315 @@ The values have been computed correctly for forward operation, however, the gene
-        ),  # part grad
-    )
-
-
-_rms_norm_bwd_p.def_abstract_eval(_rms_norm_bwd_abstract)
-
-
-#######################################
-# Top-level interface with custom vjp #
-#######################################
-
-
-@partial(jax.custom_vjp, nondiff_argnums=(2,))
-def rms_norm(x, weight, eps=1e-05):
-    output, _ = rms_norm_fwd(x, weight, eps=eps)
-    return output
-
-
-rms_norm.defvjp(rms_norm_fwd, rms_norm_bwd)
-
-
-######################
-# RMS norm with xmap #
-######################
-
-
-jax.config.update("experimental_xmap_spmd_lowering", True)
-jax.config.update("experimental_xmap_spmd_lowering_manual", True)
-
-
-def xmap_rms_norm(x, weight, *, device_count):
-    reshaped = x.reshape(device_count, x.shape[0] // device_count, *x.shape[1:])
-    xmapped = xmap(
-        rms_norm,
-        in_axes=(("x", None, None, None), (None, None)),
-        out_axes=("x", None, None, None),
-        axis_resources={"x": "x"},
-    )
-    reshaped_out = xmapped(reshaped, weight)
-    return reshaped_out.reshape(x.shape)
-
-
-########
-# Test #
-########
-
-
-import jax
-
-
-per_core_batch_size=4
-seq_len=512
-emb_dim=512
-x = jax.random.normal(
-    jax.random.key(0),
-    shape=(jax.local_device_count() * per_core_batch_size, seq_len, emb_dim),
-    dtype=jnp.bfloat16,
-)
-norm_shape = x.shape[-2:]
-weight = jnp.ones(norm_shape, dtype=jnp.bfloat16)
-
+With that we have completely defined our custom RMS norm primitive with custom_partitioning. To check for correctness we define the following loss functions: ref_loss is the reference value to compare against, while custom_p_loss uses our new primitive that implements custom_partitioning.
 
-def loss_ref(x, weight):
+```python
+def ref_loss(x, weight):
     predictions = rms_norm(x, weight)
     return -jnp.mean(predictions**2)
 
 
-ref = jax.grad(loss_ref, argnums=(0, 1))(x, weight)
+ref = jax.grad(ref_loss, argnums=(0, 1))(x, weight)
 
-
-def loss(x, weight, *, device_count):
-    predictions = xmap_rms_norm(x, weight, device_count=device_count)
+def custom_p_loss(x, weight):
+    predictions = custom_p_rms_norm(x, weight)
     return -jnp.mean(predictions**2)
+```
 
+# Check for correctness
 
+```python
 with Mesh(jax.local_devices(), ("x",)):
-
-    pjitted = pjit(
-        jax.grad(partial(loss, device_count=jax.local_device_count()), argnums=(0, 1)),
-        # Shard x by batch dimension and replicate weight on all devices.
-        in_shardings=(
-            PartitionSpec("x", None, None),
-            PartitionSpec(None, None),
-        ),
-        # Shard the output by batch dimension and replicate weight grad on all devices.
-        out_shardings=(
-            PartitionSpec("x", None, None),
-            PartitionSpec(None, None),
-        ),
-    )
-    out = pjitted(x, weight)
-
-for r, o in zip(ref, out):
-    print(jnp.allclose(r, o, atol=1e-2, rtol=1e-2))
+    def run_and_verify(loss):
+        pjitted = pjit(
```

Comment:
Would be great if this is converted to jit() as pjit() is deprecated. Can be a follow up PR too.

Reply:
I will do this in a follow up.